### PR TITLE
[Dashboard] Configure Subset of Parameters/Metrics and show Errors in Tune Dashboard

### PR DESCRIPTION
**Pull request by aannadi**
These changes give the user the ability to choose which parameters and metrics they wish to show in the table on the Tune Dashboard if there are more than 3 to choose from each. 

Screen Shot 2020-03-24 at 8 52 34 AM

The Tune Dashboard will also now show error logs associated with Tune Trials, both in the regular Tune Table with other information about the Trial and in a new Errors Tab.

Screen Shot 2020-03-24 at 8 55 24 AM
Screen Shot 2020-03-24 at 8 55 31 AM
Screen Shot 2020-03-24 at 8 55 41 AM
 

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/examples/mnist_pytorch_trainable.py [hunk 1]
```diff
@@ -45,6 +45,8 @@ def _setup(self, config):
             momentum=config.get("momentum", 0.9))
 
     def _train(self):
+        if random.random() < .1:
```
Comment (simon-mo) [inline python/ray/tune/examples/mnist_pytorch_trainable.py]: @richardliaw what's the best way to do this, add a separate example script like `dashboard_demo.py`?

File: python/ray/dashboard/dashboard.py [hunk 2]
```diff
@@ -771,12 +775,39 @@ def run(self):
                 self.collect()
             time.sleep(self._reload_interval)
 
+    def collect_errors_alt(self, job_name, df):
```
Comment (simon-mo) [inline python/ray/dashboard/dashboard.py]: can you fully spell out `alt`

File: python/ray/dashboard/client/src/api.ts [hunk 3]
```diff
@@ -217,14 +217,22 @@ export interface TuneTrial {
   training_iteration: number;
   start_time: string;
   status: string;
-  trial_id: string;
+  trial_id: string | number;
```
Comment (simon-mo) [inline python/ray/dashboard/client/src/api.ts]: can we comment on why this change is needed?

Comment (aannadi) [reply python/ray/dashboard/client/src/api.ts]: There was a change to how the trial ID's were being set such that they can now be strings or numbers. Previously, I was casting all of them to strings in the backend; however, this messes with the sorting for this attribute in the Table as it sorts them by their string value("10" < "2") rather than their numerical value.  This tweak is accompanied by a change in the backend that preserves the true type of the trial ID.

File: python/ray/tune/examples/mnist_pytorch_trainable.py (hunk 1 above)
Comment (aannadi) [reply python/ray/tune/examples/mnist_pytorch_trainable.py]: I accidentally left these changes here, never meant to commit them. Just reverted all the examples back to the original version. If we want to keep an example to test different features of the dashboard, I can make a new one.

File: python/ray/dashboard/dashboard.py (hunk 2 above)
Comment (aannadi) [reply python/ray/dashboard/dashboard.py]: Renamed to "collect_errors"

**Comment by mitchellstern** [comment]
We've added some more linting rules to the repo and adjusted formatting preferences in https://github.com/ray-project/ray/pull/7779. Please merge `master` into your PR and revert any changes to `package.json` and `package-lock.json` since all packages are now updated to their latest versions.

**Comment by mitchellstern** [comment]
By the way, please also rerun `npm install` after merging `master` to be sure you have the latest versions of the dependencies (including Prettier which has some minor formatting changes in the latest version).

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import { dashboardActions } from "../state";
+import Table from "@material-ui/core/Table";
+import TableBody from "@material-ui/core/TableBody";
+import TableCell from "@material-ui/core/TableCell";
+import TableHead from "@material-ui/core/TableHead";
+import TableRow from "@material-ui/core/TableRow";
+import NumberedLines from "../../../common/NumberedLines";
+import Link from "@material-ui/core/Link";
+import Dialog from "@material-ui/core/Dialog";
+import DialogContent from "@material-ui/core/DialogContent";
```
Comment (mitchellstern) [inline python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx]: Let's put all `@material-ui/core` imports in a single `import { ... } from "@material-ui/core"` statement and sort these imports to fix the linting errors.

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  render() {
+    const { classes, tuneInfo } = this.props;
+    const { currentError, open } = this.state;
+
+    if (tuneInfo === null || Object.keys(tuneInfo["errors"]).length === 0) {
+      return null;
+    }
+
+    return (
+      <div>
```
Comment (mitchellstern) [inline python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx]: We can use a `<React.Fragment>` here to avoid the redundant `<div>`.

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+          <DialogContent>
+            {open && (
+              <NumberedLines
+                lines={tuneInfo["errors"][currentError]["text"]
+                  .trim()
+                  .split("\n")}
+              />
+            )}
+          </DialogContent>
+        </Dialog>
```
Comment (mitchellstern) [inline python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx]: Consider using the `DialogWithTitle` component from `common/DialogWithTitle.tsx` for this. See example usage in `pages/dashboard/node-info/NodeInfo.tsx`.

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx [hunk 7]
```diff
@@ -166,9 +218,99 @@ class TuneTable extends React.Component<
...
+        metricColumns: metricColumns,
+      });
+    }
+  };
+
+  metricChoices = (metricNames: string[]) => {
+    const { metricColumns } = this.state;
+
+    return (
+      <div>
```
Comment (mitchellstern) [inline python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx]: Do we need this `<div>`?

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx [hunk 8]
```diff
@@ -180,56 +322,125 @@ class TuneTable extends React.Component<
...
+          <DialogContent>
+            {open && (
+              <NumberedLines
+                lines={tuneInfo["trial_records"][errorTrial]["error"]
+                  .trim()
+                  .split("\n")}
+              />
+            )}
+          </DialogContent>
+        </Dialog>
```
Comment (mitchellstern) [inline python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx]: Consider using the DialogWithTitle component from common/DialogWithTitle.tsx for this. See example usage in pages/dashboard/node-info/NodeInfo.tsx.

**Comment by mitchellstern** [comment]
There are a number of linting warnings / errors after our new linting PR, please address these issues as well.

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx (hunk 8 above)
Comment (aannadi) [reply python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx]: Done!

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx (hunk 7 above)
Comment (aannadi) [reply python/ray/dashboard/client/src/pages/dashboard/tune/TuneTable.tsx]: Took it out :)

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx (hunk 6 above)
Comment (aannadi) [reply python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx]: Done

File: python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx (hunk 5 above)
Comment (aannadi) [reply python/ray/dashboard/client/src/pages/dashboard/tune/TuneErrors.tsx]: Changed

**Comment by mitchellstern** [comment]
There is a Python linting error in the latest Travis run. Can you take a look @aannadi?

**Review by richardliaw — APPROVED** [review]
